[CK TILE ENGINE] Adding GEMM Preshuffle to CK Tile Engine (#2712)
ROCm/composable_kernel: tile_engine/ops/gemm_preshuffle/README.md

CK Tile Engine for GEMM Preshuffle is used to generate and run GEMM kernels with different combinations of BlockTile sizes, WarpTile sizes, WarpTile mapping for all valid pipelines, schedulers and epilogues while able to give custom datatype and Layout selections

# Kernel Configurations

# User Specific
Users can specify custom kernel configurations such as tile size, warp size, padding, pipeline, scheduler, and epilogue in the config file. This allows building only for selected configurations, significantly reducing build time.
For reference please see `./configs/user_provided_config.json`.

# Default

> The Tile engine also has a default kernel configuration for providing range of configuration parameter values, which helps users who lack kernel development experience to benchmark. For reference please see in `./configs/default_config.json`

Reviewer: ```suggestion
The Tile engine also has a default kernel configuration for providing a range of configuration parameter values, which helps users who lack kernel development experience to benchmark. For reference please see  `./configs/default_config.json`
```

> If user does not provide kernel configuration, the tile engine uses default kernel configuration to generate kernel instances and benchmark. 

Reviewer: ```suggestion
If the user doesn't provide kernel configuration, the tile engine uses the default kernel configuration to generate kernel instances and benchmarks. 
```

Is it one benchmark or several?

## Build Instructions
``` bash

> # in the root of composable kernel create build directory

Reviewer: ```suggestion
# create the `build` directory in the composable kernel root directory:
```

mkdir build && cd build
# build composable kernel

> # replace [Arch] with the appropriate architecture or leave blank and 

Reviewer: In the following command, 
* [ARCH] is the .... 

Maybe make a note to remove the []

> # replace [Datatype1;Datatype2;...] in comma separated datatypes string (possible datatypes are [fp16,fp8,bf16.bf8])

Reviewer: ```suggestion
*  [Datatype1;Datatype2;...] is a comma separated list of datatypes. Accepted datatypes are `fp16`, `fp8`, `bf16`, `bf8`.
```

> # replace [Layout1;...] in comma separated datatypes string (possible layouts are [rcr] only)

Reviewer: ```suggestion
* [Layout1;...] is a comma separated list of layouts. Only [rcr] layouts are supported.
```

is this `[rcr]` a literal, or is it any layout that follows rcr? this isn't clear

Author: rcr is a layout type

sh ../script/cmake-ck-dev.sh  ../ [Arch] -DGEMM_PRESHUFFLE_DATATYPE="[Datatype1;Datatype2]" -DGEMM_PRESHUFFLE_LAYOUT="[Layout1]" 
# generate different executable for each passed datatype
make benchmark_gemm_preshuffle_[Datatype1]_[Layout1] -j
make benchmark_gemm_preshuffle_[Datatype2]_[Layout1] -j
```
`benchmark_gemm_preshuffle_[Datatype1]_[Layout1]` will be located in the `./bin/` directory.

> `benchmark_gemm_preshuffle_[Datatype1]_[Layout1]` must be rebuilt everytime if configuration file is modified.

Reviewer: ```suggestion
`benchmark_gemm_preshuffle_[Datatype1]_[Layout1]` must be rebuilt whenever the configuration file is modified.
```

``` bash
rm -rf tile_engine/ && make benchmark_gemm_preshuffle_[Datatype1]_[Layout1] -j  # rebuild
```

## For eaxmple build for gfx942 for datatype with rcr layout
``` bash
mkdir build && cd build
sh ../script/cmake-ck-dev.sh  ../ gfx942 -DGEMM_PRESHUFFLE_DATATYPE="fp16" -DGEMM_PRESHUFFLE_LAYOUT="rcr" 
make benchmark_gemm_preshuffle_fp16_rcr -j

## benchmark_gemm inputs
```
                      -m    The value for m dimension. Default is 3840.
                      -n    The value for n dimension. Default is 4096.
                      -k    The value for k dimension. Default is 2048.
               -stride_a    The stride value for tensor A. Default is 0.
               -stride_b    The stride value for tensor B. Default is 0.
              -stride_ds    The stride value for tensor Ds. Default is 0.
               -stride_e    The stride value for tensor E. Default is 0.
                -split_k    The split value for k dimension. Default is 1.
                  -verify    The type of validation. Set to 0 for no validation, 1 for validation on CPU, or 2 for validation on GPU. Default is 1, validation on CPU, as validation on GPU is not supported.

>                     -log    Wether output kernel instance information or not. Possible values are true or false. Default is false.

Reviewer: ```suggestion
                    -log    Outputs kernel instance information when set to `true`.  Default is `false`.
```

>                  -warmup    The number of iterations before benchmark the kernel. Default is 50.

Reviewer: ```suggestion
                 -warmup    The number of iterations before benchmarking begins. Default is 50.
```

                 -repeat    The number of iterations to benchmark the kernel. Default is 100.

>                   -timer    Whether if the timer is gpu timer or not. Possible values are false or true. Default is true.

Reviewer: ```suggestion
                  -timer    The GPU timer will not be used when set to `false`. Default is `true`.
```

>                    -init    The method of tensor initialization. Set to 0 for random, to 1 for linear, or 2 for constant(1). Default is 0, random.

Reviewer: ```suggestion
                   -init    Tensor initialization method. Set to 0 for random, 1 for linear, or 2 for constant(1). Default is 0, random.
```

What does `constant(1)` mean?

>             -flush_cache    To flush cache, possible values are true or false. Default is false.

Reviewer: ```suggestion
            -flush_cache    Flushes the cache when set to `true`. Default is `false`.
```

>          -rotating_count    Number of iterations to rotate the cache. Default is 5.

Reviewer: Is this the number of iterations before the cache is rotated, or the number of iterations to use to rotate the cache?

                 -metric    Metric with which to measure kernel performance. Set to 0 for latency, 1 for tflops, or 2 for bandwidth. Default is 0, latency.

>            -csv_filename    The filename of benchmark result. Default is gemm_presshuffle

Reviewer: ```suggestion
           -csv_filename    Name of the file where benchmark results are written to. The default filename is `gemm_presshuffle`.
```

               -pipeline    The type of pipeline. Possible values are preshufflev1 or preshufflev2.

>               -scheduler    The type of scheduler. Possible values are intrawave. Default is intrawave.

Reviewer: Is `intrawave` the only possibility?

Author: for now yes

>                -epilogue    The type of epilogue. Possible values are cshuffle or default. Default is cshuffle.

Reviewer: is `cshuffle` the only possible epilogue?

Author: No, cshuffle and default are the possible epilogue

>                   -pad_m    Whether pad or not in m direction. Possible values are true or false. Default is false.

Reviewer: ```suggestion
                  -pad_m    Pads the tensor in the m direction when set to `true`. Default is `false`.
```

>                   -pad_n    Whether pad or not in n direction. Possible values are true or false. Default is false.

Reviewer: ```suggestion
                  -pad_n    Pads the tensor in the n direction when set to `true`. Default is `false`.
```

>                   -pad_k    Whether pad or not in k direction. Possible values are true or false. Default is false.

Reviewer: ```suggestion
                  -pad_k    Pads the tensor in the k direction when set to `true`. Default is `false`.
```

Note: pipeline, scheduler, epilogue, pad_m, pad_n, pad_k should be one of the options specified in user_provided_config.json 
```
Note: In `./configs/user_provided_config.json` pipeline, scheduler, epilogue, pad_m, pad_n, pad_k should be from one of the values specified above.

> 

Reviewer: These two notes sound really similar. What are we trying to say in each case?
